### PR TITLE
#Module2 #Task1 #Task2

### PR DESCRIPTION
**Pull request by TheElidanXD**
Add Module2 homework

### CODE REVIEW

File: module2/src/user-service.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    age: Joi.number().integer().min(4).max(130)
+});
+
+const validateUserSchema = (schema: Joi.ObjectSchema) => {
+    return (req: Request, res: Response, next: NextFunction) => {
+        const { error } = schema.validate(req.body, {
+            abortEarly: false,
+            allowUnknown: false
+        });
+        if (error && error.isJoi) {
```
Comment (evgenDer) [inline module2/src/user-service.ts]: error?.isJoi

File: module2/src/user-service.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+userService.listen(3000);
+router.use(express.json());
+
+router.param('id', (req, res, next, id) => {
+    const currentUser = getUserById(id);
+    if (currentUser) {
+        res.locals.user = currentUser;
+        next();
+    } else {
+        res.status(404).json({ message: `User with id ${req.params.id} not found` });
```
Comment (evgenDer) [inline module2/src/user-service.ts]: req.params.id change to id

File: module2/src/user-service.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+});
+
+router.get('/', (req, res) => {
+    res.json(userList);
+});
+
+router.get('/user/:id', (req, res) => {
```
Comment (evgenDer) [inline module2/src/user-service.ts:71]: create userRouter and move all requests with user

File: module2/src/user-service.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    res.json(res.locals.user);
+});
+
+router.post('/', validateUserSchema(userSchema), (req, res) => {
+    const newUser: User = {
+        id: uuid(),
```
Comment (evgenDer) [inline module2/src/user-service.ts:77]: I think it is better to create a model that should simulate working with database

File: module2/src/user-service.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    const currentUser = res.locals.user;
+    currentUser.isDeleted = true;
+    res.status(204).send();
+});
+
+router.get('/autoSuggestUsers/:substring/:limit', (req, res) => {
+    if (Number(req.params.limit) < 1) {
+        res.status(400).json({ message: 'Invalid limit' });
+    } else {
+        const autoSuggestUsers = getUndeletedUsers().
```
Comment (evgenDer) [inline module2/src/user-service.ts]: you shouldn't do it in a router

File: module2/src/user-service.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
+import express, { NextFunction, Request, Response } from 'express';
+import { v4 as uuid } from 'uuid';
+import { User } from './types/user';
+import Joi from 'joi';
+
+const userService = express();
+const router = express.Router();
+
+const userSchema = Joi.object().keys({
```
Comment (evgenDer) [inline module2/src/user-service.ts:9]: split all to different files - validation. controllers, routers, and e.t.c

**Review by evgenDer — CHANGES_REQUESTED** [review]

File: module2/src/user-service.ts (hunk 1 above)
Comment (TheElidanXD) [reply module2/src/user-service.ts]: Done

File: module2/src/user-service.ts (hunk 2 above)
Comment (TheElidanXD) [reply module2/src/user-service.ts]: Done

File: module2/src/user-service.ts (hunk 3 above)
Comment (TheElidanXD) [reply module2/src/user-service.ts:71]: Done in module3 HW

File: module2/src/user-service.ts (hunk 4 above)
Comment (TheElidanXD) [reply module2/src/user-service.ts:77]: done in Module 3 HW

File: module2/src/user-service.ts (hunk 5 above)
Comment (TheElidanXD) [reply module2/src/user-service.ts]: Logic moved to findUsersBySubstring function

File: module2/src/user-service.ts (hunk 6 above)
Comment (TheElidanXD) [reply module2/src/user-service.ts:9]: Done in Module 3 HW

**Pull request merged**